File: galaxy/galaxy.py
import argparse
import os
from taichi.lang.impl import grouped
import taichi as ti
import numpy as np

def from_rle(file):
    width = -1
    height = -1
    data = np.zeros((4, 4), dtype=np.uint8)
    x, y, n = 0, 0, 0
    count = 0
    with open(file, "r") as f:
        for line in f:
            if line[0] == '#':
                # skip comment lines
                continue
            elif line[0] == 'x':
                # read params
                tokens = line.split(',')
                width = int(tokens[0][4:])
                height = int(tokens[1][4:])
                data = np.zeros((height, width), dtype=np.uint8)
            else:
                # process RLE-encoded pattern
                for c in line:
                    if c.isdigit():
                        n = n * 10 + int(c)
                    else:
                        if n == 0:
                            n = 1
                        if c == 'b':
                            for j in range(x, x + n):
                                data[y][j] = 0
                            x += n
                        elif c == 'o':
                            for j in range(x, x + n):
                                data[y][j] = 1
                            x += n
                        elif c == '$':
                            x = 0
                            y += n
                        elif c == '!':
                            break
                        n = 0
    return data, width, height


parser = argparse.ArgumentParser()
parser.add_argument("--dir", type=str)
parser.add_argument("--arch", type=str)
args = parser.parse_args()

if args.arch == "cuda":
    arch = ti.cuda
elif args.arch == "x64":
    arch = ti.x64
else:
    assert False

ti.init(arch=arch)

img_size = 512
img_c = 4
N = 65536
bits = 32
n_blocks = 16
n = 30720
boundary_offset = int((N - n) / 2)

qu1 = ti.types.quant.int(1, False)

state_a = ti.field(dtype=qu1)
state_b = ti.field(dtype=qu1)
block = ti.root.pointer(ti.ij, (n_blocks, n_blocks))
block.dense(ti.ij, (N // n_blocks, N // (bits * n_blocks))).quant_array(
    ti.j, bits, max_num_bits=bits).place(state_a)
block.dense(ti.ij, (N // n_blocks, N // (bits * n_blocks))).quant_array(
    ti.j, bits, max_num_bits=bits).place(state_b)

arr = ti.ndarray(ti.f32, shape=(img_size, img_size, img_c))


@ti.kernel
def evolve_a_b():
    ti.lang.impl.get_runtime().prog.current_ast_builder().bit_vectorize()
    for i, j in state_a:
        num_active_neighbors = ti.u32(0)
        num_active_neighbors += ti.cast(state_a[i - 1, j - 1], ti.u32)
        num_active_neighbors += ti.cast(state_a[i - 1, j], ti.u32)
        num_active_neighbors += ti.cast(state_a[i - 1, j + 1], ti.u32)
        num_active_neighbors += ti.cast(state_a[i, j - 1], ti.u32)
        num_active_neighbors += ti.cast(state_a[i, j + 1], ti.u32)
        num_active_neighbors += ti.cast(state_a[i + 1, j - 1], ti.u32)
        num_active_neighbors += ti.cast(state_a[i + 1, j], ti.u32)
        num_active_neighbors += ti.cast(state_a[i + 1, j + 1], ti.u32)
        state_b[i, j] = (num_active_neighbors == 3) | ((num_active_neighbors == 2) & (state_a[i, j] == 1))

@ti.kernel
def evolve_b_a():
    ti.lang.impl.get_runtime().prog.current_ast_builder().bit_vectorize()
    for i, j in state_b:
        num_active_neighbors = ti.u32(0)
        num_active_neighbors += ti.cast(state_b[i - 1, j - 1], ti.u32)
        num_active_neighbors += ti.cast(state_b[i - 1, j], ti.u32)
        num_active_neighbors += ti.cast(state_b[i - 1, j + 1], ti.u32)
        num_active_neighbors += ti.cast(state_b[i, j - 1], ti.u32)
        num_active_neighbors += ti.cast(state_b[i, j + 1], ti.u32)
        num_active_neighbors += ti.cast(state_b[i + 1, j - 1], ti.u32)
        num_active_neighbors += ti.cast(state_b[i + 1, j], ti.u32)
        num_active_neighbors += ti.cast(state_b[i + 1, j + 1], ti.u32)
        state_a[i, j] = (num_active_neighbors == 3) | ((num_active_neighbors == 2) & (state_b[i, j] == 1))


@ti.func
def fill_pixel(scale, buffer, i, j, region_size):
    ii = i * 1.0 / img_size
    jj = j * 1.0 / img_size
    ret_val = 0.0
    if scale > 1:
        sx1, sx2, sy1, sy2 = j * scale, (j + 1) * scale, i * scale, (i + 1) * scale
        x1 = ti.cast(sx1, ti.i32)
        x2 = ti.cast(sx2, ti.i32) + 1
        y1 = ti.cast(sy1, ti.i32)
        y2 = ti.cast(sy2, ti.i32) + 1
        count = 0
        val = 0
        for mm in range(y1, y2):
            for nn in range(x1, x2):
                if mm + 0.5 > sy1 and mm + 0.5 < sy2 and nn + 0.5 > sx1 and nn + 0.5 < sx2:
                    count += 1
                    val += buffer[boundary_offset + int(n / 2) - int(region_size / 2) + mm,
                                  boundary_offset + int(n / 2) - int(region_size / 2) + nn]
        ret_val = val
    else:
        ret_val = buffer[int(boundary_offset + int(n / 2) - region_size / 2 + region_size * ii),
                   int(boundary_offset + int(n / 2) - region_size / 2  + region_size * jj)]
    return ret_val


@ti.kernel
def fill_img_a(region_size: ti.i32, arr: ti.types.ndarray()):
    scale = region_size * 1.0 / img_size
    for i, j in ti.ndrange(img_size, img_size):
        for c in range(img_c):
            arr[i, j, c] = fill_pixel(scale, state_a, i, j, region_size)

@ti.kernel
def fill_img_b(region_size: ti.i32, arr: ti.types.ndarray()):
    scale = region_size * 1.0 / img_size
    for i, j in ti.ndrange(img_size, img_size):
        for c in range(img_c):
            arr[i, j, c] = fill_pixel(scale, state_b, i, j, region_size)

@ti.kernel
def clear():
    for i in range(boundary_offset, N - boundary_offset):
        for j in range(boundary_offset, N - boundary_offset):
            state_a[i, j] = 0
            state_b[i, j] = 0

@ti.kernel
def init_from_slices(init_buffer: ti.types.ndarray(),
                     init_width: ti.i32, offset: ti.i32, rows: ti.i32):
    for i in range(boundary_offset + offset, boundary_offset + offset + rows):
        for j in range(boundary_offset, boundary_offset + init_width):
            state_a[i, j] = init_buffer[i - boundary_offset - offset, j - boundary_offset]

def save_kernels(arch, dirname):
    init_buffer, init_width, init_height = from_rle('metapixel-galaxy.rle')
    init_buffer = np.rot90(init_buffer, 3)
    
    print(init_buffer.shape)
    print(init_buffer.dtype)

    binary_filename = os.path.join(dirname, "init_buffer.bin")
    init_buffer.tofile(binary_filename)

    m = ti.aot.Module(arch)
    
    # init
    m.add_kernel(clear, template_args={})
    m.add_kernel(init_from_slices, template_args={'init_buffer': init_buffer})

    # run
    m.add_kernel(evolve_a_b, template_args={})
    m.add_kernel(evolve_b_a, template_args={})
    m.add_kernel(fill_img_a, template_args={'arr': arr})
    m.add_kernel(fill_img_b, template_args={'arr': arr})
    
    m.add_field("state_a", state_a)
    m.add_field("state_b", state_b)

    m.save(dirname, 'whatever')

if __name__ == '__main__':
    assert args.dir is not None
    save_kernels(arch=arch, dirname=args.dir)

File: galaxy/galaxy.cpp
#include <cstdlib>
#include <utility>
#include <sstream>
#include <iostream>
#include <string>
#include <assert.h>
#include <cstdio>

#include <bitset>

// C-API
#include "taichi_core_impl.h"
#include "taichi/taichi_core.h"
#include "c_api_test_utils.h"

// GUI
#include <taichi/gui/gui.h>
#include <taichi/ui/backends/vulkan/renderer.h>


constexpr int img_c = 4;
constexpr int img_size = 512;
constexpr int init_width = 30720;
constexpr int init_height = 30720;
constexpr int num_runs = 100;
constexpr int max_num_frames = 1000;
constexpr int slice_size = 1000000;
constexpr int num_rows = slice_size / init_width;
constexpr int n = 30720;

template<typename T>
std::vector<T> read_binary_from_file(const std::string& filename,
                                     size_t numel) {
    std::vector<T> result(numel);

    // Open File for Read
    FILE * pFile = fopen(filename.c_str(), "rb"); 
    if (pFile==NULL) {fputs ("File error",stderr); exit (1);}
    
    // Get File Size
    fseek (pFile , 0 , SEEK_END);
    long lSize = ftell (pFile);
    rewind (pFile);
    
    // Read and Check
    int bytes_read = fread(result.data(), sizeof(T), numel, pFile);
    if (bytes_read != lSize) {fputs ("Reading error",stderr); exit (3);}

    fclose(pFile);
    
    return result;
}

void update_sliced_array(TiRuntime runtime,
                         TiMemory memory,
                         void* buffer, 
                         int width,
                         size_t row_index,
                         size_t length,
                         size_t dtype_size) {
    void* ptr = ti_map_memory(runtime, memory);
    char* row_offset = (char*)buffer + row_index * width * dtype_size;
    size_t size_in_bytes = length * dtype_size;
    for(int i = 0; i < length * width; i++) {
        ((char*)ptr)[i] = row_offset[i];
    }

    ti_unmap_memory(runtime, memory);
    
}

static taichi::Arch get_taichi_arch(const std::string& arch_name_) {
    if(arch_name_ == "cuda") {
        return taichi::Arch::cuda;
    }

    if(arch_name_ == "x64") {
        return taichi::Arch::x64;
    }

    TI_ERROR("Unkown arch_name");
    return taichi::Arch::x64;
}
  
static TiArch get_c_api_arch(const std::string& arch_name_) {
    if(arch_name_ == "cuda") {
        return TiArch::TI_ARCH_CUDA;
    }

    if(arch_name_ == "x64") {
        return TiArch::TI_ARCH_X64;
    }
    
    TI_ERROR("Unkown arch_name");
    return TiArch::TI_ARCH_X64;
}
  
static taichi::ui::FieldSource get_field_source(const std::string& arch_name_) {
    if(arch_name_ == "cuda") {
        return taichi::ui::FieldSource::TaichiCuda;
    }

    if(arch_name_ == "x64") {
        return taichi::ui::FieldSource::TaichiX64;
    }
    
    TI_ERROR("Unkown arch_name");
    return taichi::ui::FieldSource::TaichiX64;
}


struct guiHelper {
    std::shared_ptr<taichi::ui::vulkan::Gui> gui_{nullptr};
    std::unique_ptr<taichi::ui::vulkan::Renderer> renderer{nullptr};
    GLFWwindow *window{nullptr};
    taichi::ui::SetImageInfo img_info;

    explicit guiHelper(taichi::lang::DeviceAllocation& devalloc, 
                       const std::string& arch_name) {
      glfwInit();
      glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API);
      window = glfwCreateWindow(img_size, img_size, "Taichi show", NULL, NULL);
      if (window == NULL) {
        std::cout << "Failed to create GLFW window" << std::endl;
        glfwTerminate();
      }
    
      // Create a GGUI configuration
      taichi::ui::AppConfig app_config;
      app_config.name = "TaichiSparse";
      app_config.width = img_size;
      app_config.height = img_size;
      app_config.vsync = true;
      app_config.show_window = false;
      app_config.package_path = "."; // make it flexible later
      app_config.ti_arch = get_taichi_arch(arch_name);
      app_config.is_packed_mode = true;

      // Create GUI & renderer
      renderer = std::make_unique<taichi::ui::vulkan::Renderer>();
      renderer->init(nullptr, window, app_config);

      renderer->set_background_color({0.6, 0.6, 0.6});

      gui_ = std::make_shared<taichi::ui::vulkan::Gui>(
          &renderer->app_context(), &renderer->swap_chain(), window);

      // Describe information to render the image
      taichi::ui::FieldInfo f_info;
      f_info.valid = true;
      f_info.field_type = taichi::ui::FieldType::Scalar;
      f_info.matrix_rows = 1;
      f_info.matrix_cols = 1;
      f_info.shape = { img_size, img_size, img_c };

      f_info.field_source = get_field_source(arch_name);
      f_info.dtype = taichi::lang::PrimitiveType::f32;
      f_info.snode = nullptr;
      f_info.dev_alloc = devalloc;
      
      img_info.img = std::move(f_info);
    }

    void step() {
      if (!glfwWindowShouldClose(window)) {
        // Render elements
        renderer->set_image(img_info);
        renderer->draw_frame(gui_.get());
        renderer->swap_chain().surface().present_image();
        renderer->prepare_for_next_frame();

        glfwSwapBuffers(window);
        glfwPollEvents();
      }
    }

    ~guiHelper() {
      gui_.reset();
      renderer.reset();
    }
};

static void taichi_sparse_test(const std::string& arch_name, 
                               const std::string& folder_dir) {
  TiRuntime runtime = ti_create_runtime(get_c_api_arch(arch_name));

  // Load Aot and Kernel
  TiAotModule aot_mod = ti_load_aot_module(runtime, folder_dir.c_str());

  TiKernel k_clear = ti_get_aot_module_kernel(aot_mod, "clear");
  TiKernel k_init_from_slices = ti_get_aot_module_kernel(aot_mod, "init_from_slices");
  TiKernel k_evolve_a_b = ti_get_aot_module_kernel(aot_mod, "evolve_a_b");
  TiKernel k_evolve_b_a = ti_get_aot_module_kernel(aot_mod, "evolve_b_a");
  TiKernel k_fill_img_a = ti_get_aot_module_kernel(aot_mod, "fill_img_a");
  TiKernel k_fill_img_b = ti_get_aot_module_kernel(aot_mod, "fill_img_b");
 
  /* Kernel Initialization */
  const std::vector<int> shape_2d = { num_rows, init_width };
  auto init_buffer_ = capi::utils::make_ndarray(runtime,
                                                TiDataType::TI_DATA_TYPE_U8,
                                                shape_2d.data(), 2,
                                                nullptr, 0,
                                                false /*host_read*/, false /*host_write*/
                                                );
  TiArgument init_buffer_args[4];
  init_buffer_args[0] = init_buffer_.arg_;

  TiArgument init_width_arg = {.type = TiArgumentType::TI_ARGUMENT_TYPE_I32,
                               .value = {.i32 = init_width}};
  init_buffer_args[1] = init_width_arg;
  
  // Read init_buffer from binary file -> init_buffer_val
  std::string binary_filename = folder_dir + "/init_buffer.bin";
  std::vector<char> init_buffer_val = read_binary_from_file<char>(binary_filename, init_width * init_height);

  // Initialize with slice
  ti_launch_kernel(runtime, k_clear, 0, &init_buffer_args[0]);
  for(int i = 0; i < init_height; i = i + num_rows) {
        // i
        TiArgument i_arg = {.type = TiArgumentType::TI_ARGUMENT_TYPE_I32,
                            .value = {.i32 = i}};
        init_buffer_args[2] = i_arg;
        
        // rows
        int rows = std::min(num_rows, init_height - i);
        TiArgument rows_arg = {.type = TiArgumentType::TI_ARGUMENT_TYPE_I32,
                               .value = {.i32 = rows}};
        init_buffer_args[3] = rows_arg;
        
        // init_buffer[i:i+num_rows, :] -> init_buffer_
        update_sliced_array(runtime, init_buffer_.memory_,
                            init_buffer_val.data()/* void* buffer */,
                            init_width/* int width */,
                            i/* int row_index */,
                            num_rows /* length */,
                            1 /* size_t dtype_size */);
        
        ti_launch_kernel(runtime, k_init_from_slices, 4, &init_buffer_args[0]);
  }

  /* Prepare Image for GUI*/
  TiArgument region_size_arg = {.type = TiArgumentType::TI_ARGUMENT_TYPE_I32,
                                .value = {.i32 = n}};
  
  const std::vector<int> shape_3d = { img_size, img_size, img_c };
  auto arr_ = capi::utils::make_ndarray(runtime,
                                        TiDataType::TI_DATA_TYPE_F32,
                                        shape_3d.data(), 3,
                                        nullptr, 0,
                                        false /*host_read*/, false /*host_write*/
                                        );
  TiArgument arr_args[2] = { region_size_arg, arr_.arg_ };

  Runtime* real_runtime = (Runtime *)runtime;
  taichi::lang::DeviceAllocation devalloc = devmem2devalloc(*real_runtime, arr_.memory_);
  guiHelper gui_helper(devalloc, arch_name);

  for(size_t frame = 0; frame < max_num_frames; frame++) {

      // running(state_a, state_b)
      for(size_t i = 0; i < num_runs; i++) {
        ti_launch_kernel(runtime, k_evolve_a_b, 0, &arr_args[0]);
        ti_launch_kernel(runtime, k_evolve_b_a, 0, &arr_args[0]);
      }
      
      ti_launch_kernel(runtime, k_fill_img_a, 2, &arr_args[0]);
      ti_wait(runtime);

      gui_helper.step();

      // running(state_b, state_a)
      for(size_t i = 0; i < num_runs; i++) {
        ti_launch_kernel(runtime, k_evolve_b_a, 0, &arr_args[0]);
        ti_launch_kernel(runtime, k_evolve_a_b, 0, &arr_args[0]);
      }
      ti_launch_kernel(runtime, k_fill_img_b, 2, &arr_args[0]);
      ti_wait(runtime);
      
      gui_helper.step();
  }
  

  ti_destroy_aot_module(aot_mod);
  ti_destroy_runtime(runtime);
}

int main(int argc, char *argv[]) {
    assert(argc == 3);
    std::string folder_dir = argv[1];
    std::string arch_name = argv[2];

    taichi_sparse_test(arch_name, folder_dir);

    return 0;
}
